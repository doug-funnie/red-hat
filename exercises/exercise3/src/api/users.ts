import { Address } from './address';
import { Company } from './company';
import { request } from './request';

const usersBaseUrl = 'http://jsonplaceholder.typicode.com/users';
const avatarBaseUrl = 'https://api.adorable.io/avatars';

export interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: Company;
}

export function getUsers() {
  return request<User[]>({ url: usersBaseUrl });
}

export function getUser(userId: User['id']) {
  return request<User>({ url: `${usersBaseUrl}/${userId}` });
}

export function getUserAvatarUrl(email: User['email'], size = 35) {
  return `${avatarBaseUrl}/${size}/${email}.png`;
}
