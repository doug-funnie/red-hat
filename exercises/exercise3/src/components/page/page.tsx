import glamorous, { Main } from 'glamorous';
import * as React from 'react';
import { Error } from '../error';
import { Header, height as headerHeight } from '../header';

export interface Props {
  title?: string;
  children?: React.ReactNode;
  error?: any;
  onBackButtonPress?(): void;
  onRetry?(): void;
}

const Wrapper = glamorous.div({
  margin: '0 auto',
  maxWidth: 800,
  paddingBottom: 25,
  paddingTop: 15,
  width: '100%',
});

export const Page: React.SFC<Props> = ({
  children,
  title,
  onBackButtonPress,
  error,
  onRetry,
}) => (
  <>
    <Header onBackButtonPress={onBackButtonPress} title={error ? 'Error' : title} />
    <Main marginTop={headerHeight}>
      <Wrapper>
        {!!error ? <Error onRetry={onRetry} /> : children}
      </Wrapper>
    </Main>
  </>
);
