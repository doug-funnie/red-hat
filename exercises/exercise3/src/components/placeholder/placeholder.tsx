import { keyframes } from 'glamor';
import glamorous from 'glamorous';
import * as React from 'react';
import { ThemedProps } from 'styles/theme';

export interface Props {
  isLoading: boolean;
  type: 'avatar' | 'text';
  size?: number;
  height?: number;
  width?: number;
  children?(): React.ReactNode;
}

const shimmer = keyframes({
  '0%': { opacity: 1 },
  '100%': { opacity: 0.6 },
});

type AvatarPlaceholderProps = ThemedProps<Pick<Props, 'size'>>;

export const AvatarPlaceholder = glamorous.div<AvatarPlaceholderProps>(({ size, theme }) => ({
  display: 'inline-block',
  borderRadius: '50%',
  animation: `${shimmer} 1s infinite ease-in-out alternate`,
  height: size,
  width: size,
  backgroundColor: theme.colors.grayDark,
}));

type TextPlaceholderProps = ThemedProps<Pick<Props, 'height' | 'width'>>;

export const TextPlaceholder = glamorous.span<TextPlaceholderProps>(({
  height,
  width,
  theme,
}) => ({
  display: 'inline-block',
  borderRadius: 3,
  animation: `${shimmer} 1s infinite ease-in-out alternate`,
  height,
  width,
  backgroundColor: theme.colors.grayDark,
}));

export const Placeholder: React.SFC<Props> = (props) => {
  if (!props.isLoading) {
    return props.children();
  }

  switch (props.type) {
    case 'avatar':
      return <AvatarPlaceholder size={props.size} /> as any;
    case 'text':
    return <TextPlaceholder height={props.height} width={props.width} /> as any;
  }

  return props.children();
};
