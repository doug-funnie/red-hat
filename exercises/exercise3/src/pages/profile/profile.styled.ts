import { H5 } from 'components/text';
import glamorous, {  } from 'glamorous';
import { ThemedProps } from 'styles/theme';

export const UserSheet = glamorous.div<ThemedProps>(({ theme }) => ({
  backgroundColor: theme.colors.white,
  borderRadius: 4,
  padding: 20,
}));

export const Summary = glamorous.div({
  textAlign: 'center',
});

export const AvatarContainer = glamorous.div({
  display: 'flex',
  justifyContent: 'center',
  marginBottom: 15,
});

export const InfoWrapper = glamorous.div({
  display: 'flex',
  flexWrap: 'wrap',
});

export const InfoSection = glamorous.div({
  flex: '1 1 50%',
  minWidth: 200,
});

export const InfoSectionTitle = glamorous(H5)({
  fontWeight: 700,
  marginBottom: 5,
});
