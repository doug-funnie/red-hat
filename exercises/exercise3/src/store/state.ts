import { MapDispatchToProps, MapStateToProps } from 'react-redux';
import { UserListState, UsersState } from './users';

export interface RootState {
  users: UsersState;
  userList: UserListState;
}

export const enum SortDirection {
  asc,
  desc,
}

export const enum FetchStatus {
  none = 'none',
  inProgress = 'inProgress',
  complete = 'complete',
}

const createStateKey = <K extends keyof RootState>(key: K): K => key;
export const usersStateKey = createStateKey('users');
export const userListStateKey = createStateKey('userList');

export type TypedMapStateToProps<InnerProps, OuterProps> = MapStateToProps<InnerProps, OuterProps, RootState>;
export type TypedMapDispatchToProps<DispatchProps> = MapDispatchToProps<DispatchProps, any>;
