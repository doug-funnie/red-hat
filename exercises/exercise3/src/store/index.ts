import { applyMiddleware, createStore } from 'redux';
import { combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { createRequestMiddleware } from './requestMiddleware';
import { RootState, userListStateKey, usersStateKey } from './state';
import { userListReducer, usersReducer } from './users';

export const rootReducer = combineReducers<RootState>({
  [userListStateKey]: userListReducer,
  [usersStateKey]: usersReducer,
});

export function create() {
  return createStore(
    rootReducer,
    applyMiddleware(
      thunk,
      createRequestMiddleware(),
    ),
  );
}
