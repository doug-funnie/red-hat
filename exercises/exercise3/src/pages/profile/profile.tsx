import { User } from 'api/users';
import { Avatar } from 'components/avatar/avatar';
import { Card, CardContent } from 'components/card';
import { Divider } from 'components/divider';
import { InfoSection } from 'components/infoSection';
import { Page } from 'components/page';
import { Placeholder } from 'components/placeholder/placeholder';
import { H3, H4 } from 'components/text';
import { Div } from 'glamorous';
import * as React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { bindActionCreators } from 'redux';
import { getUsersRoute } from 'routeUtil';
import { FetchStatus, TypedMapDispatchToProps, TypedMapStateToProps } from 'store/state';
import { fetchUser, selectUserById, selectUserError, selectUserFetchStatus } from 'store/users';

type OuterProps = RouteComponentProps<{ id: User['id'] }>;

interface InnerProps {
  userId: User['id'];
  user: User;
  fetchStatus: FetchStatus;
  error: any;
}

interface DispatchProps {
  fetchUser(userId: User['id']): any;
}

export type Props = InnerProps & DispatchProps & OuterProps;

export const mapStateToProps: TypedMapStateToProps<InnerProps, OuterProps> = (state, ownProps): InnerProps  => {
  const userId = ownProps.match.params.id;

  return {
    userId,
    user: selectUserById(state, userId),
    fetchStatus: selectUserFetchStatus(state, userId),
    error: selectUserError(state, userId),
  };
};

export const mapDispatchToProps: TypedMapDispatchToProps<DispatchProps> = (dispatch) => bindActionCreators({
  fetchUser,
}, dispatch);

export class Profile extends React.Component<Props> {
  public componentDidMount() {
    if (!this.props.user && this.props.fetchStatus === FetchStatus.none) {
      this.fetchUser();
    }
  }

  private fetchUser = () => {
    this.props.fetchUser(this.props.userId);
  }

  private handleBackButtonPress = () => {
    this.props.history.push(getUsersRoute());
  }

  public render() {
    const {
      user,
      fetchStatus,
      error,
    } = this.props;

    const isLoading = fetchStatus !== FetchStatus.complete;
    const title = user ? user.name : '';

    return (
      <Page
        error={isLoading ? null : error}
        onBackButtonPress={this.handleBackButtonPress}
        onRetry={this.fetchUser}
        title={title}
      >
        <Card>
          <CardContent>
            <Div display='flex' justifyContent='center' marginBottom={15}>
              <Placeholder type='avatar' size={100} isLoading={isLoading}>
                {() => <Avatar size={100} email={user.email} />}
              </Placeholder>
            </Div>
            <Div textAlign='center'>
              <Placeholder type='text' width={200} height={18} isLoading={isLoading}>
                {() => <H3>{user.name}</H3>}
              </Placeholder>
              <div>
                <Placeholder type='text' width={150} height={16} isLoading={isLoading}>
                  {() => <H4>@{user.username}</H4>}
                </Placeholder>
              </div>
            </Div>
            {!isLoading && (
              <>
                <Divider />
                <Div display='flex' flexWrap='wrap'>
                  <InfoSection
                    title='Contact Info'
                    items={[
                      { label: 'Email', value: user.email },
                      { label: 'Phone Number', value: user.phone },
                      { label: 'Website', value: user.website },
                      user.address && { label: 'City', value: user.address.city },
                    ]}
                  />
                  {!!user.company && (
                    <InfoSection
                      title='Company Info'
                      items={[
                        { label: 'Company', value: user.company.name },
                        { label: 'Description', value: user.company.catchPhrase },
                        { label: 'BS', value: user.company.bs },
                        { label: 'City', value: user.address.city },
                      ]}
                    />
                  )}
                </Div>
              </>
            )}
          </CardContent>
        </Card>
      </Page>
    );
  }
}

const ConnectedProfile = connect(mapStateToProps, mapDispatchToProps)(Profile);

export default ConnectedProfile;
