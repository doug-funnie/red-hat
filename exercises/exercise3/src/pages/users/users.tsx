import { User } from 'api/users';
import { Page } from 'components/page';
import { UserTable } from 'components/userTable';
import * as React from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { getUserProfileRoute } from 'routeUtil';
import { FetchStatus, TypedMapDispatchToProps, TypedMapStateToProps } from 'store/state';
import { fetchUsers, selectUserListFetchStatus, selectUserListSorted } from 'store/users';

type OuterProps = RouteComponentProps<{ id: string }>;

interface InnerProps {
  users: User[];
  fetchStatus: FetchStatus;
}

interface DispatchProps {
  fetchUsers(): any;
}

export type Props = InnerProps & DispatchProps & OuterProps;

export const mapStateToProps: TypedMapStateToProps<InnerProps, OuterProps> = (state) => ({
  users: selectUserListSorted(state),
  fetchStatus: selectUserListFetchStatus(state),
});

export const mapDispatchToProps: TypedMapDispatchToProps<DispatchProps> = (dispatch) => bindActionCreators(
  { fetchUsers },
  dispatch,
);

export class Users extends React.Component<Props> {
  public componentDidMount() {
    if (this.props.fetchStatus === FetchStatus.none) {
      this.props.fetchUsers();
    }
  }

  private handleUserClick = (user) => {
    this.props.history.push(getUserProfileRoute(user));
  }

  public render() {
    return (
      <Page title='Users'>
        <UserTable
          users={this.props.users}
          onUserClick={this.handleUserClick}
        />
      </Page>
    );
  }
}

const ConnectedUsers = connect(mapStateToProps, mapDispatchToProps)(Users);

export default ConnectedUsers;
