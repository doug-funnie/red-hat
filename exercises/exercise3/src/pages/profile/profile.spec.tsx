// import { User } from 'api/users';
// import { Page } from 'components/page/Page';
// import { UserTable } from 'components/userTable/userTable';
import { User } from 'api/users';
import { Page } from 'components/page';
import { Placeholder } from 'components/placeholder/placeholder';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { InfoSection } from 'pages/profile/profile.styled';
import * as React from 'react';
import { FetchStatus } from 'store/state';
import '../../../../../testUtils/addJestMatchers';
import { Profile, Props } from './profile';

const user = {
  id: 1,
  name: 'User 1',
} as User;

const props: Props = {
  user,
  userId: user.id,
  fetchStatus: FetchStatus.complete,
  error: null,
  fetchUser: jest.fn(),
  match: {} as any,
  location: {} as any,
  history: {
    push: jest.fn(),
  } as any,
};

it('fetches user if no user is available and fetchStatus is none', () => {
  shallow(<Profile {...props} user={undefined} fetchStatus={FetchStatus.none} />);
  expect(props.fetchUser).toBeCalledWith(props.userId);
});

it('sets page title to empty string while loading user', () => {
  const view = shallow(<Profile {...props} user={undefined} fetchStatus={FetchStatus.inProgress} />);
  expect(view.find(Page).props().title).toBe('');
});

it('sets page title to empty string while loading user', () => {
  const view = shallow(<Profile {...props} user={undefined} fetchStatus={FetchStatus.inProgress} />);
  expect(view.find(Page).props().title).toBe('');
});

it('does not render info sections while loading', () => {
  const view = shallow(<Profile {...props} fetchStatus={FetchStatus.inProgress} />);
  expect(view.find(InfoSection).length).toBe(0);
});

it('does not render info sections while loading', () => {
  const view = shallow(<Profile {...props} fetchStatus={FetchStatus.inProgress} />);
  expect(toJson(view.find(Placeholder))).toMatchSnapshot();
});

it('calls fetchUser on Page retry', () => {
  const view = shallow(<Profile {...props} />);
  view.find(Page).props().onRetry();
  expect(props.fetchUser).toBeCalledWith(props.userId);
});

it('navigated to user list on back', () => {
  const view = shallow(<Profile {...props} />);
  view.find(Page).props().onBackButtonPress();
  expect(props.history.push).toBeCalledWithSnapshot();
});
