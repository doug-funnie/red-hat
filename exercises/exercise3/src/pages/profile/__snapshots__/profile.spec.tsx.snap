// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not render info sections while loading 1`] = `
Array [
  <Component
    isLoading={true}
    size={100}
    type="avatar"
  />,
  <Component
    height={18}
    isLoading={true}
    type="text"
    width={200}
  />,
  <Component
    height={16}
    isLoading={true}
    type="text"
    width={150}
  />,
]
`;

exports[`navigated to user list on back 1`] = `
Array [
  Array [
    "/users",
  ],
]
`;
